C++Module: Fixing import std on enum header and using `export import std` when appropriate

M2-TE opened the pull request:
Resolves #2281.

This issue pops up when using the vulkan headers and `import std` on gcc 15.
The enums header had a leftover `#include <type_traits>` that was not properly wrapped for `import std` and would redefine the symbols within.

I've also cleaned out some headers out of `vulkan.cppm` that are already being included in `vulkan.hpp`.

sharadhr commented on vulkan/vulkan_hpp_macros.hpp:
```diff
@@ -87,7 +87,11 @@
 #endif
 
-#if defined( __cpp_lib_modules ) && !defined( VULKAN_HPP_STD_MODULE ) && defined( VULKAN_HPP_ENABLE_STD_MODULE )
+#if defined( __cpp_lib_modules ) && !defined( VULKAN_HPP_ENABLE_STD_MODULE )
+#  define VULKAN_HPP_ENABLE_STD_MODULE
+#endif
+
+#if !defined( VULKAN_HPP_STD_MODULE ) && defined( VULKAN_HPP_ENABLE_STD_MODULE )
 #  define VULKAN_HPP_STD_MODULE std.compat
 #endif
```
`__cpp_lib_modules` must be provided by the toolchain when C++23 is enabled. 

[libc++](https://github.com/llvm/llvm-project/blob/7936b6f1133516c427c478104606d9276da6f81b/libcxx/include/version#L509)

[libstdc++](https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=libstdc%2B%2B-v3/include/bits/version.h;h=b05249857d2723c5f66aff8389e4f25cb764fdff;hb=HEAD#l2280)

M2-TE replied:
> vulkan.cppm with C++ headers doesn't really make sense anyway

I don't agree, as they can very much be used and issues are easily circumvented (e.g. manually including `<compare>`. But I do agree that it would be nice to use `export import std` by default rather than headers.

The problem remains that even in C++23, CMake and co still lack the proper support for `import std`, so defaulting to it is not a good idea as of yet I believe.

M2-TE commented:
I've opted to remove the automatic defining of `VULKAN_HPP_ENABLE_STD_MODULE` for the following reasons:
1. Before the changes in this patch, it was present in the global module fragment of `vulkan.cppm`, where it seems to have never been in effect. `__cpp_lib_modules` was never being defined, even in C++23, as our CI workflows that test modules would have failed. The fact that C++23 workflows fail now that it is in the normal macro header shows this.
2. As of yet, I believe toolchains are not yet ready to automatically use `import std` when using C++23. The best example is CMake, where you have to specifically set experimental flags to even get it to work. Thus, most people would likely be unable to use the `vulkan_hpp` module without it.

sharadhr commented on vulkan/vulkan_hpp_macros.hpp:
```diff
@@ -87,7 +87,7 @@
 #  define VULKAN_HPP_SUPPORT_SPAN
 #endif
 
-#if defined( __cpp_lib_modules ) && !defined( VULKAN_HPP_STD_MODULE ) && defined( VULKAN_HPP_ENABLE_STD_MODULE )
```
I also don't follow why this feature test macro was removed.

sharadhr commented:
@M2-TE could you in summary explain how users would use the module? There are four possible scenarios now:

- C++20, `import std` disabled
- C++20, `import std` enabled (possible with extension)
- C++23, `import std` disabled 
- C++23, `import std` enabled

I believe the third configuration ought to be impossible, as a C++23-compliant compiler _must_ ship `std`.

M2-TE commented:
> It is impossible for Vulkan-Hpp to be somehow compiled a version ahead of that on client machines, as the BMI is expected to be built on said machines anyway, since it is a header-only library.

It is not impossible; the only reason some user may still use C++20 for their own code could be an incompatibility with the newer standard, unrelated to the compiler that may very much be capable of compiling for C++23.

> It is impossible for Vulkan-Hpp to be somehow compiled a version ahead of that on client machines, as the BMI is expected to be built on said machines anyway, since it is a header-only library. [...] Naturally we don't consider this as Vulkan-Hpp is source-only.

Modules with `import std` are an exception to your stated incompatibility. They require C++23 to be compiled, but only C++20 to be consumed as they do not expose anything that a normal C++20 library couldn't. CMake even has a way of handling this:

```cmake
target_compile_features(vulkan_hpp
  PRIVATE   cxx_std_23  # compile with C++23
  INTERFACE cxx_std_20) # consumable as C++20
```
So you still use the global compiler flags etc, except for a bumping the C++ version to enable `import std`.

> I don't think we should rely CMake's behaviour unless it affects internal testing

Then at least make it possible to opt-out to not completely ignore the general usability of the module under C++23.

> and instead unconditionally including the standard library headers, and then straightforwardly export import std;

Does this not go against the idea of modules in the first place; we would be unnecessarily increasing binary size by including headers and only doing the `import std` afterwards.

-----

I would propose to have an opt-out solution for C++23 in the following manner, akin to what you have shown:

### vulkan_hpp_macros.hpp

```cpp
#if !defined( VULKAN_HPP_STD_MODULE ) && defined( VULKAN_HPP_ENABLE_STD_MODULE )
#  define VULKAN_HPP_STD_MODULE std.compat
#endif
```

### vulkan.cppm

```cpp
module;

#include <version>
#if defined( __cpp_lib_modules ) && !defined( VULKAN_HPP_NO_STD_MODULE )
#  define VULKAN_HPP_ENABLE_STD_MODULE
#endif

#include <vulkan/vulkan.hpp>
// ... other includes (vulkan_hpp_macros.hpp already included via vulkan.hpp)

export module vulkan_hpp;
#if defined( VULKAN_HPP_ENABLE_STD_MODULE )
   export import VULKAN_HPP_STD_MODULE
#endif

namespace VULKAN_HPP_NAMESPACE { // ...
```

This way, users can:
1. Manually define `VULKAN_HPP_ENABLE_STD_MODULE` in case their compiler does not (bugged or otherwise)
2. Not define `VULKAN_HPP_STD_MODULE` and get a default (std.compat)
3. Opt-out by setting `VULKAN_HPP_NO_STD_MODULE` similar to other features provided by Vulkan-Hpp

sharadhr commented:
I'm happy with the opt-out! This will be more in line with the rest of Vulkan-Hpp's config options, where opt-out is the default. That said, is there a possibility to simplify `VULKAN_HPP_NO_STD_MODULE` and `VULKAN_HPP_ENABLE_STD_MODULE`? They appear to be negations of each other, could we only use one?

M2-TE commented:
Hm.. I don't think we can only use one. We need `VULKAN_HPP_NO_STD_MODULE` for opting out, so it can't be checked for enabling `import std`, only for disabling it. The "enabler" flag therefore needs to be separate, which is `VULKAN_HPP_ENABLE_STD_MODULE`. Though it's currently only necessary, because `__cpp_lib_modules` isn't defined in a C++23 module (seems to be the case for all 3 major compilers currently, should be a topic for a separate issue + discussion as to why). So we might be able to merge the flags once that issue is solved!

Current TODO:
* add proposed changes
* update README for the new opt-out flag
* add new flag to the CMake file

CI testing for std is something I'm already working on, so that will be a separate PR.

sharadhr commented:
> Does this not go against the idea of modules in the first place; we would be unnecessarily increasing binary size by including headers and only doing the import std afterwards.

Side note, a proper compiler would discard everything in the module fragment that is not explicitly `export`ed in the module interface. However this depends on compiler: MSVC does it correctly, Clang does not.

M2-TE commented:
> * add proposed changes
> * update README for the new opt-out flag
> * add new flag to the CMake file

Implemented these changes mentioned before and also added the `VULKAN_HPP_NO_STD_MODULE` to CI to prevent potential issues later down the line. From what I can see, Clang does not define `__cpp_lib_modules` at all, while gcc defines it even in C++20. The changes will also resolve #2281.

It turned out to be a little more involved than I thought at first, but it's ready for merging, so feel free to give it a final glance @sharadhr and @asuessenbach!

sharadhr commented:
> I think we should open a separate issue for discussion about that. At least, I'd like to address that in a separate PR.

Sure, although I feel it needs to be done with this PR, especially now that we have the `import export`. There's no real reason to separate it; it was written in the first place as a workaround for a compiler bug.

M2-TE commented:
@sharadhr sadly, using `export std` is not legal. Module names may only be used on `export module [...]` `import [...]` or `export import [...]` statements: [https://en.cppreference.com/w/cpp/language/modules.html](https://en.cppreference.com/w/cpp/language/modules.html).

So I believe we are stuck with using `import std` in the header for now

sharadhr commented:
Right, then we can leave it as `export import std;`.

M2-TE commented:
Placing `import std` into the `vulkan.cppm` GMF (before the `#includes`) and removing it from `vulkan.hpp` seems to break a few symbols like `std::exchange` and more on Clang. Can you test this on MSVC to see if this is compiler-specific?

sharadhr commented:
Hmm, I see something similar on MSVC:

```cpp
module;

import std;

export module Module;

export auto MyFunc()
{
	return 42ULL * std::numbers::pi;
}
```

With the warning:

```
1>R:\ConsoleApplication1\Module.ixx(3,11): warning C5202: a global module fragment can only contain preprocessor directives
```

This is only a _warning_, but the project compiles and runs nonetheless, with a consuming `main.cpp`:

```cpp
#include <iostream>

import Module;

int main()
{
    std::cout << MyFunc() << std::endl;
}
```

There is a [report on DevComm](https://developercommunity.visualstudio.com/t/Modules-Bogus-warning-C5202/10397891?sort=active) about this, but then there is also [this Reddit thread](https://www.reddit.com/r/cpp/comments/1ihbb9f/can_i_put_import_inside_the_global_module_fragment/) that claims writing `import` directly in the GMF is not allowed (but even there the wording that disallows this is... unclear).

sharadhr commented:
And @M2-TE no other comments; I appreciate the discussion and improvement.

M2-TE commented:
Likewise, thank you very much for your help and discussions @sharadhr!

asuessenbach commented:
@M2-TE and @sharadhr: It's great to have you both contributing here! It really moves forward the modules usage on Vulkan-Hpp.
Thanks a lot!

@M2-TE Would you please also adjust the newly added `ci-ubuntu-ext.yml`?

M2-TE commented:
Doesn't it all get squashed on PR merge anyways?
